### PR TITLE
Feature: Add collections methods BeProperSubsetOf / BeProperSupersetOf / BeSupersetOf

### LINKED ISSUE #2363: [API Proposal]:
### Background and motivation

Asserting proper subset and proper superset on collections

proper subset - a collection is included in a superset collection but not equivalent to the superset collection
proper superset - a collection is contains a subset collection but not equivalent to the subset collection

### API Proposal

```C#
GenericCollectionAssertions<TCollection, T, TAssertions>
{
    public AndConstraint<TAssertions> BeProperSubsetOf(IEnumerable<T> expectedSuperset, string because = "", params object[] becauseArgs);
    public AndConstraint<TAssertions> BeProperSupersetOf(IEnumerable<T> expectedSubset, string because = "", params object[] becauseArgs);
}
```

### API Usage

```C#
ISet<int> subset = new HashSet<int>{ 1, 2, 3 };
ISet<int> superset = new HashSet<int>{ 2, 3, 4 };

subset.Should().BeProperSubsetOf(new[] { 1, 2, 3 }); // will throw
subset.Should().BeProperSubsetOf(new[] { 1, 2, 3, 4 }); // will pass

superset.Should().BeProperSupersetOf(new[] { 2, 3, 4 }); // will throw
superset.Should().BeProperSupersetOf(new[] { 2, 3, 4, 5 }); // will pass
```

### Alternative Designs

```C#
subset.Should().BeSubsetOf(new[] { 1, 2, 3 }).And.NotBeEquivalentTo(new[] { 1, 2, 3 }); // for ProperSubset
new[] { 2, 3, 4 }.Should().BeSubsetOf(superset).And.NotBeEquivalentTo(superset); // for ProperSuperset
```

### Risks

_No response_

### Are you willing to help with a proof-of-concept (as PR in that or a separate repo) first and as pull-request later on?

No

### PR DESCRIPTION
**Pull request by Meir017**
resolves #2363

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

**Comment by vbreuss** [comment]
These methods come from set theory, but work on `IEnumerable<T>`. What happens when duplicate entries are provided in the subject or the expected value?
I would expect duplicate entries to be ignored, right?
What happens, e.g.
```csharp
var subject = new []{1, 1, 1, 2, 2, 3, 3};
subject.Should().BeProperSubsetOf(new []{1, 2, 3, 3, 3, 4});
```

Should we add test cases for it?

**Comment by Meir017** [comment]
@vbreuss done! thanks for the suggestion

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            var subset = new[] { 1, 1, 1, 2, 2, 3, 3, 4 };
+            var superset = new[] { 4, 4, 4, 3, 3, 2, 1, 1 };
+
+            // Act
+            Action act = () => superset.Should().BeProperSupersetOf(subset, "because we want to test the failure {0}", "message");
+
+            // Assert
+            act.Should().Throw<XunitException>().WithMessage(
+                "Expected superset to be a proper superset of {4, 3, 2, 1} because we want to test the failure message, " +
+                "but items {1, 1, 1, 2, 2, 3, 3, 4} are equivalent to the superset");
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs:81]: Should we print the non-set version of expected?
This might be useful when `object.Equals(object)` and `object.ToString()` are not aligned as they are for `int`, i.e. where `ToString()` includes more/different information than what `Equals` use to compare instances.

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 2]
```diff
@@ -664,21 +664,88 @@ public AndConstraint<TAssertions> BeNullOrEmpty(string because = "", params obje
...
+    /// </param>
+    /// <exception cref="ArgumentNullException"><paramref name="expectedProperSuperset"/> is <see langword="null"/>.</exception>
+    public AndConstraint<TAssertions> BeProperSubsetOf(IEnumerable<T> expectedProperSuperset, string because = "",
+        params object[] becauseArgs)
+    {
+        AssertBeSubsetOf(expectedProperSuperset, "proper subset", because, becauseArgs);
+
+        ISet<T> expectedItems = expectedProperSuperset.ConvertOrCastToSet();
+
+        if (expectedItems.Intersect(Subject).Count() == expectedItems.Count)
```
[the remark this answers is not in the document]
Comment (Meir017) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:712]: hmm, maybe we should create an `ISet` and an `ICollection` for the `Subject` and `expected` values?

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs (hunk 1 above)
Comment (Meir017) [reply Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs:81]: hmm, maybe we should create an `ISet` and an `ICollection` for the `Subject` and `expected` values?
then we could print both versions of the values, both the unique values and the raw values

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 3]
```diff
@@ -3262,6 +3298,62 @@ public AndConstraint<TAssertions> StartWith(T element, string because = "", para
...
+                    Execute.Assertion
+                        .BecauseOf(because, becauseArgs)
+                        .FailWith("Expected {context:collection} {0} to " + containmentType + " {1}{reason}.",
+                            Subject, expectedObjects.Single());
+                }
+            }
+        }
+    }
+
+    internal void AssertBeSubsetOf(IEnumerable<T> expectedSuperset, string subsetType, string because = "",
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:3337]: 🔧 This should have a `private` scope

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 4]
```diff
@@ -664,21 +664,88 @@ public AndConstraint<TAssertions> BeNullOrEmpty(string because = "", params obje
...
-            .Given(subject => subject.Except(expectedSuperset))
-            .ForCondition(excessItems => !excessItems.Any())
-            .FailWith("but items {0} are not part of the superset.", excessItems => excessItems)
-            .Then
-            .ClearExpectation();
+        return new AndConstraint<TAssertions>((TAssertions)this);
+    }
+
+    /// <summary>
+    /// Asserts that the collection is a superset of the <paramref name="expectedSubset" />.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:673]: 🤔 I think it's worth the trouble to clarify what all these new methods mean. E.g. to make it totally clear what a proper subset is versus a "normal" subset.

File: docs/_pages/collections.md [hunk 5]
```diff
@@ -41,6 +41,9 @@ collection.Should().EndWith(8);
 collection.Should().EndWith(new[] { 5, 8 });
 
 collection.Should().BeSubsetOf(new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, });
+collection.Should().BeProperSubsetOf(new[] { 1, 2, 5, 6, 7, 8 });
```
Comment (dennisdoomen) [inline docs/_pages/collections.md:44]: 🤔 I would some comments here to explain the difference (again) between a normal subset/superset and a proper one.

File: docs/_pages/releases.md [hunk 6]
```diff
@@ -12,6 +12,10 @@ sidebar:
 ### What's new
 
 * Introduced a new assembly-level attribute that you can use to initialize Fluent Assertions before the first assertion - [#2292](https://github.com/fluentassertions/fluentassertions/pull/2292)
+* Introduced new collection assertions methods - [#2432](https://github.com/fluentassertions/fluentassertions/pull/2432)
+  * `BeProperSubsetOf`
+  * `BeProperSupersetOf`
+  * `BeSupersetOf`
```
Comment (dennisdoomen) [inline docs/_pages/releases.md:18]:
```suggestion
* Introduced new collection assertions methods `BeProperSubsetOf`, `BeProperSupersetOf` and `BeSupersetOf - [#2432](https://github.com/fluentassertions/fluentassertions/pull/2432)
  * `BeSupersetOf`
```

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 7]
```diff
@@ -664,21 +664,88 @@ public AndConstraint<TAssertions> BeNullOrEmpty(string because = "", params obje
...
+    {
+        AssertBeSubsetOf(expectedProperSuperset, "proper subset", because, becauseArgs);
+
+        ISet<T> expectedItems = expectedProperSuperset.ConvertOrCastToSet();
+
+        if (expectedItems.Intersect(Subject).Count() == expectedItems.Count)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected {context:collection} to be a proper subset of {0}{reason}, but items {1} are equivalent to the subset",
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:716]: 🤔 Can we make the last part of the failure clearer?

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 8]
```diff
@@ -3262,6 +3298,62 @@ public AndConstraint<TAssertions> StartWith(T element, string because = "", para
         return StartWith(new[] { element }, ObjectExtensions.GetComparer<T>(), because, becauseArgs);
     }
 
+    internal void AssertContainment(IEnumerable<T> expected, string containmentType, string because = "", params object[] becauseArgs)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:3301]: 🙃 We prefer to keep private methods just below the last usage.
🔧 Make it `private`

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+/// <content>
+/// The BeProperSubsetOf specs.
+/// </content>
+public partial class CollectionAssertionSpecs
+{
+    public class BeProperSubsetOf
+    {
+        [Fact]
+        public void When_collection_is_proper_subset_of_a_specified_collection_it_should_not_throw()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs:16]: 🔧 We prefer to avoid using terms like _when_ and _should_.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 10]
```diff
@@ -651,6 +651,7 @@ public AndConstraint<TAssertions> BeNullOrEmpty(string because = "", params obje
 
     /// <summary>
     /// Asserts that the collection is a subset of the <paramref name="expectedSuperset" />.
+    /// If <paramref name="expectedSuperset" /> is equivalent to the subject then it's still a valid superset since it contains all of the elements from the subject.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:654]: ⛏️ It is _it's_, if you mean _it is_. It is _its_ if you mean possession.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Not all previous comments have been addressed either.

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs (hunk 9 above)
Comment (Meir017) [reply Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs:16]: I see many assertions use this styling (ex: `CollectionAssertionSpecs.BeNullOrEmpty`), 
what would be a better description for this?

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs (hunk 7 above)
Comment (Meir017) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs]: improved the error message

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs (hunk 1 above)
Comment (Meir017) [reply Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSupersetOf.cs:81]: done!

File: docs/_pages/collections.md [hunk 11]
```diff
@@ -40,7 +40,10 @@ collection.Should().StartWith(new[] { 1, 2 });
 collection.Should().EndWith(8);
 collection.Should().EndWith(new[] { 5, 8 });
 
-collection.Should().BeSubsetOf(new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, });
+collection.Should().BeSubsetOf(new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, }); // It should be contained or equivalent to the supetset.
+collection.Should().BeProperSubsetOf(new[] { 1, 2, 5, 6, 7, 8 }); // It should be contained but not equivalent to the superset.
+collection.Should().BeSupersetOf(new[] { 1, 2, 5, 8 }); // It should contain or be equivalent to the subset.
+collection.Should().BeProperSupersetOf(new[] { 1, 5, 2 }); // It should contain but not be equivalent to the subset.
```
Comment (dennisdoomen) [inline docs/_pages/collections.md:46]: 🔧This won't fit on a single. See https://fluentassertions.com/collections/. Using comments on top of the examples, or separate section will probably work better. Also note that you can run the documentation website locally from the `docs` folder.

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs (hunk 9 above)
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeProperSubsetOf.cs:16]: That is because we started changing the style a year or two ago. We now prefer a more factual style. 

In this case, `A_collection_with_only_items_of_a_superset_but_misses_some_is_a_proper_subset`. It'll take a bit more time to come up with but has really helped us and contributors understand the subtleties of our test cases.

File: docs/_pages/collections.md (hunk 11 above)
Comment (dennisdoomen) [inline docs/_pages/collections.md:46]: 🔧This won't fit on a single. See https://fluentassertions.com/collections/. Using comments on top of the examples, or separate section will probably work better. Also note that you can run the documentation website locally from the `docs` folder.

**Review by dennisdoomen — APPROVED** [review]
Nice work

**Comment by ITaluone** [comment]
@Meir017 FYI we usually use [rebasing](https://github.com/fluentassertions/fluentassertions/blob/develop/CONTRIBUTING.md#dos-and-donts) instead of merging in changes from develop ;)

**Comment by dennisdoomen** [comment]
And there are some Qodana issues to address.

**Comment by Meir017** [comment]
@dennisdoomen done

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeSupersetOf.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+using Xunit.Sdk;
+
+namespace FluentAssertions.Specs.Collections;
+
+/// <content>
+/// The BeSupersetOf specs.
+/// </content>
+public partial class CollectionAssertionSpecs
+{
+    public class BeSupersetOf
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeSupersetOf.cs:13]: We're missing test on how `BeSupersetOf` should behave when `Subject` is either `null` or empty.

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeSupersetOf.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            // Arrange
+            var collection = new[] { 1, 2, 3 };
+
+            // Act
+            Action act = () => collection.Should().BeSupersetOf(new int[0]);
+
+            // Assert
+            act.Should().Throw<ArgumentException>().WithMessage(
+                "Cannot verify containment against an empty collection*");
+        }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.BeSupersetOf.cs:83]: I'm sorry, I didn't think about this case when I suggested forwarding `BeProperSupersetOf` to `Contain`.

The empty set is a valid set and `{ 1, 2, 3 }` is a super set of `{}`.

I think all the set theoretic methods should handle empty sets.

**Review by jnyrup — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Collections/GenericCollectionAssertions.cs [hunk 14]
```diff
@@ -666,21 +667,95 @@ public AndConstraint<TAssertions> BeNullOrEmpty(string because = "", params obje
...
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected {context:collection} to be a proper subset of {0}{reason}, but items {1} are equivalent to the superset {2}",
+                    expectedItems, Subject, expectedProperSuperset);
+        }
+
+        return new AndConstraint<TAssertions>((TAssertions)this);
+    }
+
+    /// <summary>
+    /// Asserts that the collection is a proper subset of the <paramref name="expectedProperSubset" />.
```
[the remark this answers is not in the document]
Comment (ITaluone) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:731]: Don't forget to rename the parameter as well in every place

```diff
- IEnumerable<T> expectedProperSubset
+ IEnumerable<T> expectedProperSuperset
```

[the remark this answers is not in the document]
Comment (jnyrup) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:731]: > Don't forget to rename the parameter as well in every place
> 
> ```diff
> - IEnumerable<T> expectedProperSubset
> + IEnumerable<T> expectedProperSuperset
> ```

They are correct as they are.
`BeSupersetOf(IEnumerable<T> expectedSubset)`
The `Subject` is the superset, the expectation is the subset.

[the remark this answers is not in the document]
Comment (ITaluone) [reply Src/FluentAssertions/Collections/GenericCollectionAssertions.cs:731]: ou.. you are totally right.. sorry about this :D

**Comment by dennisdoomen** [comment]
@Meir017 Would you mind rebasing on `develop` and then address the last remarks?

**Comment by dennisdoomen** [comment]
This PR can no longer be reopened because its base branch (\develop\) has been deleted. I've ported this work (crediting @Meir017) into a new PR against \main\: #3271. Thanks for the original contribution!
